sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java

```diff
@@ -0,0 +1,165 @@
+   *
+   * @param sqlNode SqlNode representing a call to a function
+   * @return Granularity as intended by the function call
+   * @throws ParseException SqlNode cannot be converted a granularity
+   */
+  public static Granularity convertSqlNodeToGranularity(SqlNode sqlNode) throws ParseException
+  {
+
+    final String genericParseFailedMessageFormatString = "Unable to parse the granularity from %s. Expected HOUR, DAY, MONTH, YEAR, ALL TIME, FLOOR function or %s function";
+
+    if (!(sqlNode instanceof SqlCall)) {
+      throw new ParseException(StringUtils.format(
+          genericParseFailedMessageFormatString,
+          sqlNode.toString(),
+          TimeFloorOperatorConversion.SQL_FUNCTION_NAME
+      ));
+    }
+    SqlCall sqlCall = (SqlCall) sqlNode;
+
+    String operatorName = sqlCall.getOperator().getName();
+
+    Preconditions.checkArgument(
+        "FLOOR".equalsIgnoreCase(operatorName)
+        || TimeFloorOperatorConversion.SQL_FUNCTION_NAME.equalsIgnoreCase(operatorName),
+        StringUtils.format(
+            "PARTITIONED BY clause can only parse FLOOR and %s functions.",
+            TimeFloorOperatorConversion.SQL_FUNCTION_NAME
+        )
+    );
+
+    List<SqlNode> operandList = sqlCall.getOperandList();
+    Preconditions.checkArgument(
+        operandList.size() == 2,
+        StringUtils.format("Invalid number of arguments passed to %s in PARTIITONED BY clause", operatorName)
+    );
+
+
+    // Check if the first argument passed in the floor function is __time
+    SqlNode timeOperandSqlNode = operandList.get(0);
+    Preconditions.checkArgument(
+        timeOperandSqlNode.getKind().equals(SqlKind.IDENTIFIER),
+        StringUtils.format("First argument to %s in PARTITIONED BY clause can only be __time", operatorName)
+    );
+    SqlIdentifier timeOperandSqlIdentifier = (SqlIdentifier) timeOperandSqlNode;
+    Preconditions.checkArgument(
+        timeOperandSqlIdentifier.getSimple().equals(ColumnHolder.TIME_COLUMN_NAME),
+        StringUtils.format("First argument to %s in PARTITIONED BY clause can only be __time", operatorName)
+    );
+
+    // If the floor function is of form TIME_FLOOR(__time, 'PT1H')
+    if (operatorName.equalsIgnoreCase(TimeFloorOperatorConversion.SQL_FUNCTION_NAME)) {
+      SqlNode granularitySqlNode = operandList.get(1);
+      Preconditions.checkArgument(
+          granularitySqlNode.getKind().equals(SqlKind.LITERAL),
+          "Second argument to TIME_FLOOR in PARTITIONED BY clause should be string representing a period"
+      );
+      String granularityString = SqlLiteral.unchain(granularitySqlNode).toValue();
+      Period period;
+      try {
+        period = new Period(granularityString);
+      }
+      catch (IllegalArgumentException e) {
+        throw new ParseException(StringUtils.format("Unable to create period from %s", granularitySqlNode.toString()));
```

Comment:
```suggestion
        throw new ParseException(StringUtils.format("Unable to create period from %s", granularityString));
```

Reply:
I used the other form because then I won't need to explicitly put the string in quotes. Plus it represents exactly what the user typed, not the one curated for our use.